contracts/0.8.25/utils/UpgradeTemplateV3.sol

```diff
@@ -0,0 +1,639 @@
+    bytes32 internal constant REQUEST_BURN_MY_STETH_ROLE = keccak256("REQUEST_BURN_MY_STETH_ROLE");
+    // PauseUntilWithRoles
+    bytes32 internal constant RESUME_ROLE = keccak256("RESUME_ROLE");
+    bytes32 internal constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    // OracleReportSanityChecker
+    bytes32 internal constant ALL_LIMITS_MANAGER_ROLE = keccak256("ALL_LIMITS_MANAGER_ROLE");
+    bytes32 public constant EXITED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE =
+        keccak256("EXITED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE");
+    bytes32 public constant APPEARED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE =
+        keccak256("APPEARED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE");
+    bytes32 public constant ANNUAL_BALANCE_INCREASE_LIMIT_MANAGER_ROLE =
+        keccak256("ANNUAL_BALANCE_INCREASE_LIMIT_MANAGER_ROLE");
+    bytes32 public constant SHARE_RATE_DEVIATION_LIMIT_MANAGER_ROLE =
+        keccak256("SHARE_RATE_DEVIATION_LIMIT_MANAGER_ROLE");
+    bytes32 public constant MAX_VALIDATOR_EXIT_REQUESTS_PER_REPORT_ROLE =
+        keccak256("MAX_VALIDATOR_EXIT_REQUESTS_PER_REPORT_ROLE");
+    bytes32 public constant MAX_ITEMS_PER_EXTRA_DATA_TRANSACTION_ROLE =
+        keccak256("MAX_ITEMS_PER_EXTRA_DATA_TRANSACTION_ROLE");
+    bytes32 public constant MAX_NODE_OPERATORS_PER_EXTRA_DATA_ITEM_ROLE =
+        keccak256("MAX_NODE_OPERATORS_PER_EXTRA_DATA_ITEM_ROLE");
+    bytes32 public constant REQUEST_TIMESTAMP_MARGIN_MANAGER_ROLE = keccak256("REQUEST_TIMESTAMP_MARGIN_MANAGER_ROLE");
+    bytes32 public constant MAX_POSITIVE_TOKEN_REBASE_MANAGER_ROLE =
+        keccak256("MAX_POSITIVE_TOKEN_REBASE_MANAGER_ROLE");
+    bytes32 public constant SECOND_OPINION_MANAGER_ROLE =
+        keccak256("SECOND_OPINION_MANAGER_ROLE");
+    bytes32 public constant INITIAL_SLASHING_AND_PENALTIES_MANAGER_ROLE =
+        keccak256("INITIAL_SLASHING_AND_PENALTIES_MANAGER_ROLE");
+    // StakingRouter
+    bytes32 public constant REPORT_REWARDS_MINTED_ROLE = keccak256("REPORT_REWARDS_MINTED_ROLE");
+    // VaultHub
+    bytes32 public constant VAULT_MASTER_ROLE = keccak256("Vaults.VaultHub.VaultMasterRole");
+    bytes32 public constant VAULT_REGISTRY_ROLE = keccak256("Vaults.VaultHub.VaultRegistryRole");
+
+    //
+    // Values for checks to compare with or other
+    //
+
+    uint256 internal constant EXPECTED_FINAL_LIDO_VERSION = 3;
+    uint256 internal constant EXPECTED_FINAL_ACCOUNTING_ORACLE_VERSION = 3;
+
+    //
+    // Constants
+    //
+
+    // Timestamp since startUpgrade() and finishUpgrade() revert with Expired()
+    // This behavior is introduced to disarm the template if the upgrade voting creation or enactment didn't
+    // happen in proper time period
+    uint256 public constant EXPIRE_SINCE_INCLUSIVE = 1754006400; // 2025-08-01 00:00:00 UTC
+
+    // Initial value of _upgradeBlockNumber
+    uint256 internal constant UPGRADE_NOT_STARTED = 0;
+
+    //
+    // Immutables
+    //
+    bool immutable public ALLOW_NON_SINGLE_BLOCK_UPGRADE;
```

Comment:
:thinking: probably it's better to put the hack outside the contract (and can be resolved opportunistically within HardHat)

Reply:
Last resort: EIP-7702 hack to batch it (TBA ~1-2 weeks)